[![Code Ocean Logo](images/CO_logo_135x72.png)](http://codeocean.com/product)

<hr>

# STAR RNA-Seq alignment

Fast alignment of RNA-seq data to DNA references with or without known splice junctions. See [STAR Manual](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf) for more details.

## Input data

Fastq files:

In the **data** directory, single or paired end sequencing files in fastq format, optionally gzipped.

Genome Index Directory: A genome index generated by STAR Index, including the following files:

- Index
    - chrLength.txt
    - chrName.txt
    - chrNameLength.txt
    - chrStart.txt
    - exonGeTrInfo.tab
    - exonInfo.tab
    - geneInfo.tab
    - Genome
    - genomeParameters.txt
    - SA
    - SAindex
    - sjdbInfo.txt
    - sjdbList.fromGTF.out.tab
    - transcriptInfo.tab

If a reference directory is not supplied, it is inferred by the location of SAindex
## App Panel Parameters

Threads
- Number of threads used by STAR. If not specified, will use all threads available [default: None]

STAR Genome Directory
- STAR genome. Must be preprocessed using STAR index tool. See documentation for more details [default: ../data/Index]

Pattern fwd
- Pattern for forward sequence in paired end reads. i.e., sample_S1_L001_R1_001.fastq.gz would be "_R1_001.fastq.gz". Note, in this example, if you have multiple lanes for a single sample that you wish to process together you will need to concatenate them together before processing with STAR. [default: _R1.fastq.gz]

Pattern rev
- Pattern for reverse sequence in paired end reads. The sample prefix (i.e, sample_S1_L001 in the previous example) will be used with this pattern to determine the matching paired end fastq. If this is not found, STAR will proceed in single end mode. [default: _R2.fastq.gz]

Read files command
- Unzip command to read fastq files. **IF YOU HAVE UNCOMPRESSED DATA THIS MUST BE CHANGED TO cat!** 
 [default: zcat]

Sort Type 
- **Unsorted** outputs unsorted files (these can be processed downstream by software expecting name sorted reads such as HTseq) 
- **SortedByCoordinate** outputs sorted and indexed .bams. [default]
- **Unsorted SortedByCoordinate** outputs both unsorted and sorted .bams. 

Quant Mode
- **TranscriptomeSAM** output alignments translated into transcript coordinates in the Aligned.toTranscriptome.out.bam file. see [STAR Manual](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf) for more details.
- **GeneCounts** STAR will count number reads per gene while mapping
- **TranscriptomeSAM GeneCounts** Output both
- **-** Output nothing [default]

Output Unmapped Reads
- **Within** will output to .bam file
- **Within KeepPairs** will (redundantly) record unmapped mate
for each alignment, and, in case of unsorted output, keep it adjacent to its mapped mate (this only
affects multi-mapping reads).
- **FastX** will output unmapped reads into Unmapped.out.mate1.
- **None** output no unmapped reads [default]

Chimeric output (for fusion calling)
- **Junctions**: Chimeric.out.junction
- **SeparateSAMold**: output old SAM into separate Chimeric.out.sam file
- **WithinBAM HardClip**: hard clip supplemental chimeric alignments [default]
- **WithinBAM SoftClip**: soft clip supplemental chimeric alignments

Two pass alignment 
- **Per sample**: STAR will perform the 1st pass mapping, then it will automatically extract junctions, insert them into the genome index, and, finally, re-map
all reads in the 2nd mapping pass
- **Multi-sample**: STAR will perform 1st pass mapping on ALL samples. On the second pass, STAR will include .SJ.out.tab files for all previous samples
- **None**: STAR will only perform 1st pass mapping. [default]

*Note*: Multi-sample may be more suited for a Pipeline since the first pass can run in a separate instance for each sample, then be followed by a collection step for the junction outputs and then the second pass can also run on separate instances. 

In general, for larger samples, it may be better to run in a pipeline since each sample can use its own instance. 

Shared Memory
- Determines whether samples share the same genome in RAM or reload the genome each time. This is incompatible with Per Sample two pass and the second pass for multi sample two pass. Generally, this is preferred for small samples (~3 million reads) since there is overhead each time the genome is loaded. [default: False]

RAM reserved for BAM sort (GB)
- Only used with shared memory. Determines the amount of RAM to reserve for bam sorting. Should be at least 10. [default: 10]

## Outputs

These can vary depending on output parameters, but by default: 

**Aligned.sortedByCoord.out.bam**: Output alignments

**Log.final.out, Log.out, Log.progress.out**: Logs generated by STAR. Log.out contains parameter defaults and which parameters are overwritten. 

**SJ.out.tab**: Splice junctions. Format is specified in [STAR Manual](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf).

<hr>

[Code Ocean](https://codeocean.com/) is a cloud-based computational platform that aims to make it easy for researchers to share, discover, and run code.<br /><br />
[![Code Ocean Logo](images/CO_logo_68x36.png)](https://www.codeocean.com)

